### PR TITLE
Introduce `exitDuration` to Tooltip for mouse pointer devices

### LINKED ISSUE #136586: Assigning showDuration in tooltip applies same duration to both long press and tap released and mouse pointer exits
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Create a tooltip, assign it "message" and "triggerMode => tap" and test it on a phone and a computer.
The default values ​​are shown, 1.5 seconds for finger tapping and 0.1 seconds for mouse hover.

Then also assign "showDuration" of 5 seconds for example. Test the tooltip on your phone and computer.

### Expected results

I would expect showDuration to only modify the finger tap, and not the mouse hover since with the mouse it is the user who has control over how long they want to show the tooltip. Or create two different showDurations, one for finger tapping and one for mouse.

The actions are different especially for tooltips with longer messages. On the phone the user clicks, reads and in the meantime the tooltip disappears after a certain number of seconds. While on the computer, the user hovers over the icon and holds it until he is finished reading (because there is no finger stopping him from reading), however as soon as he releases the icon with the mouse the countdown starts and, as shown in the video, the tooltip remains even when it scrolls.

### Actual results

Instead showDuration modifies both actions, and if you assign showDuration both tapping and hovering will be changed

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: ListView(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Tooltip(
              message: "message message message message message message message message message message message message message message message",
              showDuration: Duration(seconds: 5),
              triggerMode: TooltipTriggerMode.tap,
              child: Icon(
                Icons.air_sharp
              ),
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/83022092/87e4d8b7-a78e-45f1-8614-50a9d7f53e44

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.13.6, on macOS 14.0 23A344 darwin-arm64, locale it-IT)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] VS Code (version 1.82.2)
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by JaspervanRiet**
This PR introduces a new property `exitDuration` to Tooltip, the counterpart to `waitDuration`. The need for this is shown by #136586. This changes the behaviour of `showDuration` on mouse pointer devices. This is because the use cases for the current behaviour on touch screen devices vs mouse pointer devices is fundamentally different.

<details>
<summary>Demo: tooltip with showDuration set</summary>

Tooltip disappears after 100 ms when moving away the mouse. Tooltip will not disappear when hovered.

https://github.com/flutter/flutter/assets/5138348/81d36dc9-78e0-4723-a84b-2552843ee181

</details>

Currently, when `showDuration` is set, this adjusts the time it takes for the tooltip to hide _after_ a mouse pointer has left the tooltip. This is not the same use case as its effect on touch screen devices, where it dictates how long the tooltip stays on screen after a long press. That is needed because the tooltip takes up screen space and there is not an intuitive way to hide it, whereas when using a mouse users expect to simply have to hover somewhere else. Having the tooltip stay around will look broken.

Thus, this PR splits the two use cases. `showDuration` no longer affects mouse pointer devices at all*. There is a precedent for such mouse pointer-only behaviour in `waitDuration`. Instead, I have split up the two use cases and created the new property `exitDuration`, which will still allow for tweaking the time it takes for the tooltip to hide after the user has moved their mouse pointer somewhere else.

*Note: Should `showDuration` affect [this line](https://github.com/flutter/flutter/blob/e33d4b86270e3c012ba13d68d6e90f2eabc4912b/packages/flutter/lib/src/material/tooltip.dart#L610)?

Fixes #136586.

Note: I noticed that when I made the change, no tests were broken. Hopefully, the tests added here help that in the future. I also noticed that in the _existing_ tests, the `waitDuration` tests contain assertions that implicate that it is the role of `waitDuration` to change this behaviour, but that's not currently (nor in the new behaviour) true, so I have fixed those tests.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/tooltip.dart [hunk 1]
```diff
@@ -301,13 +302,18 @@ class Tooltip extends StatefulWidget {
...
   /// The length of time that the tooltip will be shown after a long press is
   /// released (if triggerMode is [TooltipTriggerMode.longPress]) or a tap is
-  /// released (if triggerMode is [TooltipTriggerMode.tap]) or mouse pointer
-  /// exits the widget.
+  /// released (if triggerMode is [TooltipTriggerMode.tap]). This property
+  /// does not affect mouse pointer devices.
   ///
-  /// Defaults to 1.5 seconds for long press and tap released or 0.1 seconds
-  /// for mouse pointer exits the widget.
+  /// Defaults to 1.5 seconds for long press and tap released
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/tooltip.dart:308]: Maybe add a "See also:" section here that points to the `exitDuration`?

File: packages/flutter/lib/src/material/tooltip.dart [hunk 2]
```diff
@@ -301,13 +302,18 @@ class Tooltip extends StatefulWidget {
...
   final Duration? showDuration;
 
+  /// The length of time that a pointer must have stopped hovering over a
+  /// tooltip's widget before the tooltip will be hidden.
+  ///
+  /// Defaults to 100 milliseconds.
+  final Duration? exitDuration;
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/tooltip.dart:315]: Same here, adding a "See also:" for `showDuration`.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Thanks for the PR (and fixing the tests!)

**Comment by gspencergoog** [comment]
Looks like `examples/api/test/material/tooltip/tooltip.1_test.dart` is failing.

File: packages/flutter/test/material/tooltip_test.dart [hunk 3]
```diff
@@ -2083,6 +2083,104 @@ void main() {
...
+    addTearDown(gesture.removePointer);
+    await gesture.moveTo(tester.getCenter(find.byTooltip(tooltipText)));
+    await tester.pump(const Duration(seconds: 1));
+    expect(find.text(tooltipText), findsOneWidget, reason: 'Tooltip should be visible when hovered.');
+
+    await gesture.moveTo(Offset.zero);
+    await tester.pump(const Duration(milliseconds: 100));
+    await tester.pumpAndSettle();
+    expect(find.text(tooltipText), findsOneWidget,
+      reason: 'Tooltip should be wait before exitDuration expires before being hidden');
```
Comment (cbracken) [inline packages/flutter/test/material/tooltip_test.dart:2177]:
```suggestion
      reason: 'Tooltip should wait until exitDuration expires before being hidden');
```

**Review by cbracken — APPROVED** [review]
lgtm with one small nit.

**Comment by LongCatIsLooong** [comment]
> *Note: Should showDuration affect [this line](https://github.com/flutter/flutter/blob/e33d4b86270e3c012ba13d68d6e90f2eabc4912b/packages/flutter/lib/src/material/tooltip.dart#L610)?

I think showDuration makes more sense there since that line only schedules a dismiss timer when there is no hover pointer keeping the tooltip visible.

File: packages/flutter/test/material/tooltip_test.dart (hunk 3 above)
Comment (JaspervanRiet) [reply packages/flutter/test/material/tooltip_test.dart]: Done 👍

File: packages/flutter/lib/src/material/tooltip.dart (hunk 1 above)
Comment (JaspervanRiet) [reply packages/flutter/lib/src/material/tooltip.dart:308]: Done 👍

File: packages/flutter/lib/src/material/tooltip.dart (hunk 2 above)
Comment (JaspervanRiet) [reply packages/flutter/lib/src/material/tooltip.dart:325]: Done 👍

**Comment by JaspervanRiet** [comment]
> Looks like examples/api/test/material/tooltip/tooltip.1_test.dart is failing.

@gspencergoog I have updated the test. It was coupled to the old behaviour.

**Comment by JaspervanRiet** [comment]
@gspencergoog Looks like it's failing checks now. Do I need to do anything?

**Comment by gspencergoog** [comment]
I don't think it has anything to do with your change.  I've merged changes from the base branch, but the tree is broken at the moment anyhow, so we'll wait until it is green and try it again.

**Pull request merged**